Editorial review: Document web app manifest localization (#44609)
mdn/content: files/en-us/web/progressive_web_apps/manifest/reference/lang/index.md

---
title: lang
slug: Web/Progressive_web_apps/Manifest/Reference/lang
page-type: web-manifest-member
browser-compat: manifests.webapp.lang
sidebar: pwasidebar
---

> The `lang` manifest member is used to specify a default language for your web application, which will be used unless overriden by a different `lang` value found in a [`*_localized`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/*_localized) member.

Reviewer: I'm nearing the end of my review, and I realize that this first doc that I read in the review is confusing:

> overriden by a different `lang` value found in a [`*_localized`](/en-US/docs/Web/Progressive_web_apps/Manifest/Reference/*_localized) member.

If it specifies the default language for my web application, what sense does it make to override?

Author: See previous comments.

> 

Reviewer: Is there a default `lang` assumed?

Reviewer 2: I think including the manifest-level `dir` and `lang` as part of this conversation can be confusing. 
In the Chromium implementation, the language selection is independent of the manifest-level `lang` specification, and is solely based on the user's browser language.

Author: I've updated this. The chosen language has nothing to do with `lang`. When localized members are included (e.g. `name_localized`), the browser chooses the variant that matches the user's browser language setting.

If no variant matches the user's browser language setting, the value in the non-localized member is used instead (e.g. `name`). I've explained this in the `*_localized` reference and the Howto.

## Syntax

```json-nolint
"lang": "en-US"
"lang": "ja"
"lang": "uk"
```

### Values

a {{glossary("BCP 47 language tag")}} representing the app's default language.

## Examples

### Basic usage

> in this example, the app's default `lang` is `en-US`, and its default `dir` is `ltr`. Its default `name` is "Color picker", but we've also specified localized variants in the `name_localized` member. Users that have their primary language preference set to German (`de`), French (`fr`), or Arabic (`ar`) will see the app's name displayed in a more suitable language for them.

Reviewer: Capitalise sentence. "for them" doesn't feel quite right. I'd do it this way.

```suggestion
In this example, the app's default `lang` is `en-US` and default `dir` is `ltr`.
The default `name` is "Color picker", but we have also defined localized variants in the `name_localized` member. Users that have their primary language preference set to German (`de`), French (`fr`), or Arabic (`ar`) will see the app's name displayed in their selected language.
```

> The French variant is specified as French Canadian (`fr-CA`) in its `lang` property, with a direction (`dir`) of `ltr`, while the Arabic variant has its `dir` specified as `rtl`. The German variant doesn't need its `lang` or `dir` specified, so its value is a string containing the localized text.

Reviewer: This doesn't quite work because the French Canadian version doesn't need its direction set either.

Author: I've chatted to @kyerebo about the dir and lang members' involvement in manifest localization, and ... they are confusingly not involved at all. I've therefore made quite a few changes to the articles.

In this particular case (and on the `dir` page), I've cut the example right down, removing all this stuff.

```json
{
  "lang": "en-US",
  "dir": "ltr",
  "name": "Color Picker",
  "name_localized": {
    "de": "Farbwähler",

>     "fr": { "value": "Sélecteur de Couleur", "lang": "fr-CA", "dir": "ltr" },

Reviewer: How does this work - i.e this is fr, but you also have a `lang` that is "fr-CA". However I know `lang` sets the default language.

My **guess** is that this should omit `lang` and fix up the key.

```
    "fr-CA": { "value": "Sélecteur de Couleur", "dir": "ltr" },
```

Further, in the intro it says "will see the app's name displayed in a more suitable language for them."

I would like to understand the fallback here. If my language is set to `fr`, will it choose  `"fr-CA"` or will it choose English ? Can I declare both "fr" and "fr-CA", and so on. The example could show a case or two.

Reviewer: Also see https://github.com/mdn/content/pull/44609/changes#r3517444725

Reviewer: Definitely see https://github.com/mdn/content/pull/44609/changes#r3517481036 :-)

Reviewer 2: I can't speak to the standards approach to this, but for the Chromium implementation, the localized member's `lang` field, if specified, doesn't currently have an effect on what the user sees. 
The primary criteria for localization selection is if there exists a language tag that matches or partly matches the user's current browser language setting. 

So if we had something like:
```
"fr": { "value": "Color Selector, "lang": "en" },
```
The end result for a browser set to French is that we'd see `Color Selector`. 
The `lang` field is effectively acting as a "heads up, this value is _this_ language rather than the one specified by the language tag."

Author: Again, I've cut this right down. In other places, I've made the text more understandable.

In the particular case of the `lang` property being different from the main object property key, I've removed the `fr-CA` example and used a much more obvious one; see the Howto guide's `short_name_localized` example.
